tsl/src/continuous_aggs/refresh.c

```diff
@@ -374,3 +374,54 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 			 "continuous aggregate \"%s\" is already up-to-date",
 			 NameStr(cagg->data.user_view_name));
 }
+
+/*
+ * Refresh all continuous aggregates on a hypertable.
+ *
+ * The refreshing happens in a single transaction. For this to work correctly,
+ * there must be no new invalidations written in the refreshed region during
+ * the refresh. Therefore, the caller is responsible for proper locking to
+ * ensure there are no invalidations (INSERTs, DELETEs, etc.). For instance,
+ * exclusively locking the hypertable or the individual chunks covered by the
+ * region would work.
+ */
+void
+continuous_agg_refresh_all(const Hypertable *ht, int64 start, int64 end)
+{
+	Catalog *catalog = ts_catalog_get();
+	List *caggs = ts_continuous_aggs_find_by_raw_table_id(ht->fd.id);
+	Dimension *dim = hyperspace_get_open_dimension(ht->space, 0);
+	InternalTimeRange refresh_window = {
+		.type = ts_dimension_get_partition_type(dim),
+		.start = start,
+		.end = end,
+	};
+	ListCell *lc;
+
+	Assert(list_length(caggs) > 0);
+	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
```

Comment:
It would be great to have a concurrent tests where we this refresh locking is actually tested in deterministic way. I believe @mkindahl recently added support for functionality which allows to control: pause and resume on breakpoints, for eg. after reaching this lock. Not suggesting to do it now, unless it is easy to do.

Reply:
It is in commit https://github.com/timescale/timescaledb/pull/2150/commits/f03a967def67a0b6fa8a360a1b8ec2f16a7fc2a4 if you want to have a look.

Reply:
There are some tests already, and some for the old materializer that I want to move over. But I agree that we should definitely improve testing for this. We are also taking pretty heavy locks right now so there's room to optimize and improve this in the future, e.g., to support concurrent refreshes in disjoint regions.